prog/prog.go

```diff
@@ -130,68 +117,123 @@ func (a *Arg) Size() uintptr {
+	ArgCommon
+	Res   Arg     // reference to arg which we use
+	OpDiv uintptr // divide result (executed before OpAdd)
+	OpAdd uintptr // add to result
+	Val   uintptr // value used if Res is nil
+}
+
+// This argument denotes syscall return value.
+type ReturnArg struct {
+	ArgCommon
+}
+
+// Returns inner arg for pointer args.
+func InnerArg(arg Arg) Arg {
+	if t, ok := arg.Type().(*sys.PtrType); ok {
+		if a, ok := arg.(*PointerArg); ok {
+			if a.Res == nil {
+				if !t.Optional() {
+					panic(fmt.Sprintf("non-optional pointer is nil\narg: %+v\ntype: %+v", a, t))
+				}
+				return nil
+			} else {
+				return InnerArg(a.Res)
+			}
+		}
+		return nil // *ConstArg.
 	}
-	r.Uses[arg] = true
-	return arg
+	return arg // Not a pointer.
 }
 
-func dataArg(t sys.Type, data []byte) *Arg {
-	return &Arg{Type: t, Kind: ArgData, Data: append([]byte{}, data...)}
+func encodeValue(value, size uintptr, bigEndian bool) uintptr {
+	if !bigEndian {
+		return value
+	}
+	switch size {
+	case 2:
+		return uintptr(swap16(uint16(value)))
+	case 4:
+		return uintptr(swap32(uint32(value)))
+	case 8:
+		return uintptr(swap64(uint64(value)))
+	default:
+		panic(fmt.Sprintf("bad size %v for value %v", size, value))
+	}
 }
 
-func pointerArg(t sys.Type, page uintptr, off int, npages uintptr, obj *Arg) *Arg {
-	return &Arg{Type: t, Kind: ArgPointer, AddrPage: page, AddrOffset: off, AddrPagesNum: npages, Res: obj}
+func constArg(t sys.Type, v uintptr) Arg {
+	return &ConstArg{ArgCommon: ArgCommon{typ: t}, Val: v}
 }
 
-func pageSizeArg(t sys.Type, npages uintptr, off int) *Arg {
-	return &Arg{Type: t, Kind: ArgPageSize, AddrPage: npages, AddrOffset: off}
+func resultArg(t sys.Type, r Arg, v uintptr) Arg {
+	arg := ResultArg{ArgCommon: ArgCommon{typ: t}, Res: r, Val: v}
+	var ret Arg = &arg
+	if r == nil {
+		return ret
+	}
+	if *r.Uses() == nil {
+		*r.Uses() = make(map[Arg]bool)
+	}
+	if (*r.Uses())[ret] {
+		panic("already used")
+	}
+	(*r.Uses())[ret] = true
+	return ret
 }
 
-func groupArg(t sys.Type, inner []*Arg) *Arg {
-	return &Arg{Type: t, Kind: ArgGroup, Inner: inner}
+func dataArg(t sys.Type, data []byte) Arg {
+	return &DataArg{ArgCommon: ArgCommon{typ: t}, Data: append([]byte{}, data...)}
 }
 
-func unionArg(t sys.Type, opt *Arg, typ sys.Type) *Arg {
-	return &Arg{Type: t, Kind: ArgUnion, Option: opt, OptionType: typ}
+func pointerArg(t sys.Type, page uintptr, off int, npages uintptr, obj Arg) Arg {
+	return &PointerArg{ArgCommon: ArgCommon{typ: t}, PageIndex: page, PageOffset: off, PagesNum: npages, Res: obj}
 }
 
-func returnArg(t sys.Type) *Arg {
+func groupArg(t sys.Type, inner []Arg) Arg {
+	return &GroupArg{ArgCommon: ArgCommon{typ: t}, Inner: inner}
+}
+
+func unionArg(t sys.Type, opt Arg, typ sys.Type) Arg {
+	return &UnionArg{ArgCommon: ArgCommon{typ: t}, Option: opt, OptionType: typ}
+}
+
+func returnArg(t sys.Type) Arg {
 	if t != nil {
-		return &Arg{Type: t, Kind: ArgReturn, Val: t.Default()}
+		return &ReturnArg{ArgCommon: ArgCommon{typ: t}}
```

Comment:
You dropped the `Val: t.Default()` part. If it's not needed (is it?), then you don't need the if as well.